### PR TITLE
Function that just builds and executes the query without returning the result set

### PR DESCRIPTION
**Pull request by cithukyaw**
Right now, `find()`, `query()` and `execute()` are returning array of result set. Sometimes, it needs to fetch the query result row by row explicitly to avoid the expensive array processing. So, added a new method `Model->runQuery()` that just builds and executes the query and returns the data source object to fetch the result manually, for example,

```
$db = $MyModel->runQuery();
if ($db) {
  // $db->numRows returns the number of rows fetched
  while ($row = $db->fetchResult()) {
    // do something with $row
  }
}
```

`DboSource->read()` is updated by internally calling a new method `DboSource->getQueryStatement()`.

Refs: #6426

### CODE REVIEW

File: lib/Cake/Model/Datasource/DboSource.php [hunk 1]
```diff
@@ -1157,6 +1105,83 @@ public function read(Model $Model, $queryData = array(), $recursive = null) {
 	}
 
 /**
+ * Build SQL statement with the primary model, plus hasOne and belongsTo associationsand
```
Comment (markstory) [inline lib/Cake/Model/Datasource/DboSource.php]: missing a space before 'and'

File: lib/Cake/Model/Model.php [hunk 2]
```diff
@@ -3359,6 +3359,26 @@ public function resetAssociations() {
...
 
 /**
+ * This gets the association arrays according to the recursive level set
+ * @return array Array of associations or empty array
+ */
+	public function getAssociationsByRecursive() {
+		if ($this->recursive === -1) {
+			// Primary model data only, no joins.
+			$associations = array();
+
```
Comment (markstory) [inline lib/Cake/Model/Model.php]: You could return here.

File: lib/Cake/Model/Model.php [hunk 3]
```diff
@@ -3438,6 +3458,38 @@ public function query($sql) {
 	}
 
 /**
+ * Builds and execute the query. Return the PDOStatement to retrieve the result explicitly using `fetchResult()`
+ *
+ * @param array $query Option fields (conditions / fields / joins / limit / offset / order / page / group / callbacks)
+ * @param int $recursive Number of levels of association
+ * @return mixed PDOStatement if query executes with no problem, true as the result of a successful, false on error
+ */
+	public function exeQuery($query = array(), $recursive = null) {
```
Comment (markstory) [inline lib/Cake/Model/Model.php]: `execQuery()`, or `runQuery()`?

File: lib/Cake/Model/Model.php [hunk 4]
```diff
@@ -3438,6 +3458,38 @@ public function query($sql) {
 	}
 
 /**
+ * Builds and execute the query. Return the PDOStatement to retrieve the result explicitly using `fetchResult()`
+ *
+ * @param array $query Option fields (conditions / fields / joins / limit / offset / order / page / group / callbacks)
+ * @param int $recursive Number of levels of association
+ * @return mixed PDOStatement if query executes with no problem, true as the result of a successful, false on error
```
Comment (markstory) [inline lib/Cake/Model/Model.php]: The return seems to be a Datasource, not a PDOStatement instance.

**Comment by robertpustulka** [comment]
I like the idea, but `exeQuery` doesn't feel right. `rawQuery`? `runQuery`?

File: lib/Cake/Model/Model.php (hunk 3 above)
Comment (cithukyaw) [inline lib/Cake/Model/Model.php]: I prefer `runQuery()`. I will update it.

File: lib/Cake/Model/Model.php (hunk 4 above)
Comment (cithukyaw) [inline lib/Cake/Model/Model.php]: Yes, my bad.

File: lib/Cake/Model/Model.php (hunk 2 above)
Comment (cithukyaw) [inline lib/Cake/Model/Model.php]: Okay. I will rebase.

**Comment by cithukyaw** [comment]
I renamed `exeQuery()` to `runQuery()`.

File: lib/Cake/Model/Datasource/DboSource.php [hunk 5]
```diff
@@ -1157,6 +1105,83 @@ public function read(Model $Model, $queryData = array(), $recursive = null) {
 	}
 
 /**
+ * Build SQL statement with the primary model, plus hasOne and belongsTo associations and
+ * and return it in order to use with the query execution methods
+ *
+ * @param Model $Model A Model object that the query is for.
+ * @param array $queryData An array of queryData information containing keys similar to Model::find(), but it is passed by reference.
```
Comment (dereuromark) [inline lib/Cake/Model/Datasource/DboSource.php]: Should this be `&$queryData`?

File: lib/Cake/Model/Datasource/DboSource.php [hunk 6]
```diff
@@ -1157,6 +1105,83 @@ public function read(Model $Model, $queryData = array(), $recursive = null) {
...
+ * @param int $recursive Number of levels of association
+ * @return string The built SQL statement
+ */
+	public function getQueryStatement(Model $Model, &$queryData = null, $recursive = null) {
+		$queryData = is_null($queryData) ? array() : $queryData;
```
Comment (dereuromark) [inline lib/Cake/Model/Datasource/DboSource.php]: === null

File: lib/Cake/Model/Model.php [hunk 7]
```diff
@@ -3438,6 +3456,38 @@ public function query($sql) {
...
+ * @return mixed DataSource object if query executes with no problem, null on failure
+ */
+	public function runQuery($query = array(), $recursive = null) {
+		$db = clone $this->getDataSource();
+		if (!$db) {
+			return null;
+		}
+
+		$queryData = $this->buildQuery('all', $query);
+		if (is_null($queryData)) {
```
Comment (dereuromark) [inline lib/Cake/Model/Model.php]: === null

File: lib/Cake/Model/Model.php [hunk 8]
```diff
@@ -3438,6 +3456,38 @@ public function query($sql) {
 	}
 
 /**
+ * Builds and execute the query. Return DataSource object to retrieve the result explicitly using `fetchResult()`
+ *
+ * @param array $query Option fields (conditions / fields / joins / limit / offset / order / page / group / callbacks)
+ * @param int $recursive Number of levels of association
```
Comment (dereuromark) [inline lib/Cake/Model/Model.php]: int|null ?

File: lib/Cake/Model/Datasource/DboSource.php (hunk 6 above)
Comment (cithukyaw) [inline lib/Cake/Model/Datasource/DboSource.php]: I think `is_null` and `=== null` has no difference.

File: lib/Cake/Model/Model.php (hunk 7 above)
Comment (cithukyaw) [inline lib/Cake/Model/Model.php]: I think `is_null` and `=== null` has no difference. However, there was [a benchmark](https://hakre.wordpress.com/2011/03/29/php-is_null-vs-null/) that mentioned `is_null` is slower. It is an old post back in 2011 though.

Comment (ADmad) [inline lib/Cake/Model/Model.php]: @cithukyaw It's more about code consistency rather than speed. Throughout the codebase we use `=== null`.

**Comment by markstory** [comment]
I'm going to close this. While I appreciate the effort you have put into this I feel that this approach has a few problems that will prove to be problematic in the long term:

* The state of result set is stored in the Connection. This means you cannot execute one query, and then execute another.
* The connection containing the resultset for a longer period of time feels like we are missing a better abstraction like statement object.

**Comment by cithukyaw** [comment]
Did you mean `$db = clone $this->getDataSource()` and the result set are being stored in `$db` Couldn't we fix it?

**Comment by markstory** [comment]
Cloning the datasource feels like a workaround for a missing statement object to me.

**Pull request closed without merging**